File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& arr) {
        int n = arr.size();
        vector<vector<int>> ans;
        sort(arr.begin(),arr.end());
        for(int i=0;i<n;i++){
           if(i>0 &&  arr[i] == arr[i-1]) continue;
            int a = arr[i];
            int l = i+1, r = n-1;
            while(l<r){
                int ThreeSum = a + arr[l] + arr[r];
                if(ThreeSum > 0){
                    r--;
                }
                else if(ThreeSum < 0){
                    l++;
                }
                else{
                    ans.push_back({a,arr[l],arr[r]});
                    l++;
                    while(arr[l] == arr[l-1] && l<r) l++;
                }
            }
        }       
        return ans;
    }
};

File: 0015-3sum/0015-3sum.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        n = len(nums)
        ans = []
        nums.sort()

        for i, a in enumerate(nums):
            if(i > 0 and nums[i] == nums[i-1]) :
                continue

            l = i+1
            r = n-1
            while(l < r):
                threeSum = a + nums[l] + nums[r]

                if threeSum > 0:
                    r -= 1
                elif threeSum < 0:
                    l += 1
                else:
                    ans.append([a, nums[l], nums[r]])
                    l += 1
                    while nums[l] == nums[l-1] and l < r:
                        l += 1
        return ans

        